prime-dotnet-webapi/Services/EnrolleePaperSubmissionService.cs

```diff
@@ -264,5 +264,126 @@ public async Task FinalizeSubmissionAsync(int enrolleeId)
+
+        public async Task<bool> LinkEnrolmentToPaperEnrolmentAsync(int enrolmentId, int paperEnrolmentId, bool isConfirmed = false)
+        {
+            var enrollee = await _context.Enrollees
+                .Where(
+                    e => e.Id == enrolmentId
+                    && _context.EnrolleeLinkedEnrolment.Any(link => link.EnrolleeId == e.Id)
+                )
+                .AnyAsync();
+
+            var paperEnrollee = await _context.Enrollees
+                .Where(
+                    pe => pe.GPID.StartsWith(PaperGpidPrefix)
+                    && pe.Id == paperEnrolmentId
+                    && _context.EnrolleeLinkedEnrolment.Any(
+                        link => link.PaperEnrolleeId == pe.Id
+                        )
+                )
+                .AnyAsync();
+
+            if (paperEnrollee)
+            {
+                return false;
+            }
+
+            var enrolleeLinkedEnrolment = await _context.EnrolleeLinkedEnrolment
+                .Where(ele => ele.EnrolleeId == enrolmentId)
+                .SingleOrDefaultAsync();
+
+            enrolleeLinkedEnrolment.PaperEnrolleeId = paperEnrolmentId;
+            enrolleeLinkedEnrolment.EnrolmentLinkDate = DateTime.Now;
+            enrolleeLinkedEnrolment.Confirmed = isConfirmed;
+
+            await _context.SaveChangesAsync();
+
+            return true;
+        }
+
+        public async Task CreateInitialLinkAsync(int enrolleeId, string userProvidedGpid)
+        {
+            var enrolleeLinkedEnrolment = await _context.EnrolleeLinkedEnrolment
+                .Where(ele => ele.EnrolleeId == enrolleeId)
+                .SingleOrDefaultAsync();
+
+            if (enrolleeLinkedEnrolment == null)
+            {
+                var newLinkedEnrolment = new EnrolleeLinkedEnrolment
+                {
+                    EnrolleeId = enrolleeId,
+                    UserProvidedGpid = userProvidedGpid,
+                };
+
+                _context.Add(newLinkedEnrolment);
+            }
+            else
+            {
+                enrolleeLinkedEnrolment.UserProvidedGpid = userProvidedGpid;
+            }
+
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task UpdateLinkedGpidAsync(int enrolleeId, string userUpdatedGpid)
+        {
+            var enrolleeLinkedEnrolment = await _context.EnrolleeLinkedEnrolment
+                .Where(ele => ele.EnrolleeId == enrolleeId)
+                .SingleOrDefaultAsync();
+
+            enrolleeLinkedEnrolment.UserProvidedGpid = userUpdatedGpid;
+            _context.SaveChanges();
+        }
+
+        public async Task<string> GetLinkedGpidAsync(int enrolleeId)
+        {
+            var value = await _context.EnrolleeLinkedEnrolment
+            .AsNoTracking()
+            .Where(ele => ele.EnrolleeId == enrolleeId)
+            .Select(ele => new
+            {
+                ele.UserProvidedGpid
+            })
+            .DecompileAsync()
```

Comment:
`.DecompileAsync()` is used when you want your linq statements that include _computed_ properties to run against the database. 
So Enrollees, for instance, have `CurrentStatus`. It's not a table that EF can just join to, its actually the following code:

```c#
[Computed]  
public EnrolmentStatus CurrentStatus
{
    get => EnrolmentStatuses
        .OrderByDescending(s => s.StatusDate)
        .ThenByDescending(s => s.Id)
        .FirstOrDefault();
}
```

so if you wanted to do something like `_context.Enrollees.Where(e => e.CurrentStatus.StatusCode == StatusType.Active)` you would have to add `.DecompileAsync` at the end so that EF can translate that into SQL.

So you don't need it here, since you aren't using any `[Computed]` properties.

Reply:
Removed